### PR TITLE
SIMSBIOHUB-10: pending review status for related datasets

### PR DESCRIPTION
**Pull request by anissa-agahchen**
# Overview

## Links to Jira tickets

[ **SIMSBIOHUB-10**](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-10)

## Description of relevant changes

- Added supplementary data that reflects if the related dataset has documents pending review
- Front end changes to show pending review status when applicable

### CODE REVIEW

File: api/src/services/security-service.ts [hunk 1]
```diff
@@ -272,4 +272,23 @@ export class SecurityService extends DBService {
...
+   * @memberof SecurityService
+   */
+  async isDatasetPendingReview(datasetId: string): Promise<boolean> {
+    const artifactIds = (await this.artifactService.getArtifactsByDatasetId(datasetId)).map((item) => item.artifact_id);
+
+    const artifactSecurityRules = await Promise.all(
+      artifactIds.map(async (artifactId) => await this.isArtifactPendingReview(artifactId))
+    );
+
+    const isPendingReview = artifactSecurityRules.find((item) => item === true) ? true : false;
```
Comment (curtisupshall) [inline api/src/services/security-service.ts]: I would change this to
```javascript
const isPendingReview = artifactSecurityRules.includes(true)
```

File: api/src/services/submission-service.ts [hunk 2]
```diff
@@ -419,13 +425,15 @@ export class SubmissionService extends DBService {
       return [];
     }
 
-    return result[0].value.map((relatedProject: any) => {
+    const relatedDatasets = result[0].value.map((relatedProject: any) => {
```
Comment (curtisupshall) [inline api/src/services/submission-service.ts]: I would revert this change - guessing this is a testing artifact

**Review by curtisupshall — DISMISSED** [review]
Code looks good! Two minor nitpicks

File: api/src/paths/dwc/submission/{datasetId}/related.ts [hunk 3]
```diff
@@ -80,12 +90,24 @@ export function getRelatedDatasetsByDatasetId(): RequestHandler {
...
+        datasets.map(async (dataset) => {
+          const isDatasetPendingReview = await securityService.isDatasetPendingReview(dataset.datasetId);
+
+          return {
+            ...dataset,
+            supplementaryData: { isPendingReview: isDatasetPendingReview }
+          };
+        })
+      );
+
```
Comment (NickPhura) [inline api/src/paths/dwc/submission/{datasetId}/related.ts]: Discussed with Anissa, but never wrote down. So adding here just in case:

Its not clear, but I think we can assume based on the frontend logic, that we don't want the API to reveal any information about the artifact if it is in a pending state, to non-admin users.

So this supplementary data section, should really only return `isPendingReview` if the calling user is an admin, and return an empty object if they are a non admin. The openapi and frontend types/interfaces would need to be updated accordingly of course, to make `isPendignReview` optional.

This same logic should also apply to our other Artifact endpoint that returns the `SECURE`, `UNSECURE`, `PENDING` statuses. If the user is an admin, no change. But if they are a non admin, we should really only return `SECURE` or `UNSECURE` (if the status is PENDING, we would return SECURE instead). This exact logic is handled this way on the frontend, but really should be a backend task, and the frontend should just display the status without having to check what role the user has. So the backend will get slightly more complicated, but the frontend will get simpler, and overall the API response data is more secure.

**Review by al-rosenthal — APPROVED** [review]
ran locally, looks good 🌞

**Review by curtisupshall — APPROVED** [review]
Code looks good 🌳
